# Applied programming in Python



## Lesson 1: Introduction to Python

Let's get to know the main terms, learn the basic syntax of Python and find out what is the *dynamic programming* and *typing*. We will consider the basic data structures (*list, tuple, dict, set*) and learn how to perform operations on them. We will find out how to write loops (*for, while*) and conditional constructs (*if, elif, else*), as well as discuss why we need *break* and *continue*.

https://www.youtube.com/watch?v=7f2FiEVL87o&ab_channel=karpov.courses

<iframe width="1000" height="600" src="https://www.youtube.com/embed/7f2FiEVL87o" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

### Basic mathematical operations

In Python, the following mathematical operations can be performed using the standard mathematical operators:

- Addition: +
- Subtraction: -
- Multiplication: *
- Division: /
- Modulus: %
- Exponentiation: **
- Floor division: //

In Python, the modulus operator (%) is used to find the remainder after division. For example:


In Python, the modulus operator (%) is used to find the remainder after division. For example:

In [ ]:
# Find the remainder after dividing 8 by 3
remainder = 8 % 3

# The remainder is 2, so the value of `remainder` is 2
print(remainder)


In this code, 8 % 3 is evaluated to find the remainder after dividing 8 by 3. This is 2, so the value of remainder is 2.

Floor division (//) is used to find the quotient of a division operation, but it rounds the result down to the nearest integer. For example:

In [ ]:
# Find the quotient of 7 divided by 3
quotient = 7 // 3

# The quotient is rounded down to the nearest integer, so the value of `quotient` is 2
print(quotient)


In this code, 7 // 3 is evaluated to find the quotient of 7 divided by 3. This is 2.33, but since we are using floor division, the result is rounded down to the nearest integer, so the value of quotient is 2.

Additionally, Python has a number of built-in functions for performing more advanced mathematical operations, such as trigonometric functions (sin(), cos(), etc.), logarithms (log(), log10(), etc.), and more. These functions are typically found in the math module, which must be imported in order to use them. For example:

In [ ]:
# Import the math module
import math

# Use the math module to calculate the sine of 30 degrees
sine = math.sin(math.radians(30))

### Types of variables in Python

https://youtu.be/KudqIBpyHu4

<iframe width="1000" height="600" src="https://www.youtube.com/embed/KudqIBpyHu4" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

In Python, variables can have different types. Some of the main types of variables in Python are:

1. None (null)
2. Boolean (True, False)
3. Numeric (int, float, complex)
4. Sequence (list, tuple, range)
5. Text Sequence (str)
6. Binary Sequence (bytes, bytearray, memoryview)
7. Set (set, frozenset)
8. Mapping (dict)
  
Here are some examples of variables with different types in Python:\

<img src="https://static.javatpoint.com/python/images/python-data-types.png" width="400">

In [ ]:
# Integer
x = 1

# Float
y = 2.5

# String
z = "hello"

# Boolean
flag = True

# List
numbers = [1, 2, 3]

# Tuple
coordinates = (1, 2)

# Dictionary
person = {"name": "John", "age": 30}


The type of a variable in Python can be determined using the type() function. For example, the following code prints the types of the variables defined above:

In [ ]:
print(type(x)) # <class 'int'>
print(type(y)) # <class 'float'>
print(type(z)) # <class 'str'>
print(type(flag)) # <class 'bool'>
print(type(numbers)) # <class 'list'>
print(type(coordinates)) # <class 'tuple'>
print(type(person)) # <class 'dict'>

#### Numeric types

In Python, there are several different types that can be used to represent numbers. These include:

- **int**: This is the most basic type of number, and represents a simple integer value (a whole number without a decimal point). For example, 5 and -3 are both examples of int values.
- **float**: This type represents a number with a decimal point, allowing for fractional values. For example, 3.14 and -0.01 are both examples of float values.
- **complex:** This type represents a complex number, which is a number with both a real and imaginary component. For example, 2 + 3j is a complex number with a real component of 2 and an imaginary component of 3.
- **bool**: This type represents a boolean value, which is either True or False.

In [1]:
# Define an integer
x = 5

# Define a float
y = 3.14

# Define a complex number
z = 3 + 2j

# Print the type of each variable
print(type(x)) # int
print(type(y)) # float
print(type(z)) # complex

<class 'int'>
<class 'float'>
<class 'complex'>


#### Mutable and immutable objects

In Python, a variable is a named location in memory that is used to store a value. The type of a variable refers to the kind of value that it can hold, such as an integer, a string, or a list.

There are two main types of variables in Python: mutable and immutable.

Mutable variables are those that can be changed after they are created. For example, a list is a mutable type because you can add, remove, or modify the items in a list.

Immutable variables, on the other hand, are those that cannot be changed after they are created. For example, a string is an immutable type because once you create a string, you cannot change the characters in it.

<img src="https://notesformsc.org/wp-content/uploads/2021/07/Python-Mutable-Immutable-2.png" width="400">

In [2]:
# Define a list (mutable)
my_list = [1, 2, 3]

# Modify the list
my_list.append(4)

# Print the modified list
print(my_list) # [1, 2, 3, 4]


# Define a string (immutable)
my_string = "Hello"

# Try to modify the string
my_string[0] = "H"

# This will cause an error, because strings are immutable


[1, 2, 3, 4]


TypeError: 'str' object does not support item assignment

#### Sequence types of variables

In Python, a sequence is a data type that represents a group of elements that can be accessed by their position, or index, in the sequence. There are several built-in sequence types in Python, including:

- **str**: This is the string type, and it represents a sequence of characters. For example, "Hello" is a string that contains the characters 'H', 'e', 'l', 'l', and 'o'.
- **list**: This is the list type, and it represents a sequence of values. For example, [1, 2, 3] is a list that contains the values 1, 2, and 3.
- **tuple**: This is the tuple type, and it represents a sequence of values that cannot be changed. For example, (1, 2, 3) is a tuple that contains the values 1, 2, and 3.

In [3]:
# Define a string
my_string = "Hello"

# Define a list
my_list = [1, 2, 3]

# Define a tuple
my_tuple = (1, 2, 3)

# Access elements in the string
print(my_string[0]) # 'H'

# Access elements in the list
print(my_list[1]) # 2

# Access elements in the tuple
print(my_tuple[2]) # 3


H
2
3


Tuple is using () and list is using [].

#### Operators

In Python, operators are special symbols that are used to perform operations on values, such as arithmetic operations or assignment. There are several types of operators in Python, including:

- **Arithmetic operators**: These operators are used to perform basic arithmetic operations, such as addition, subtraction, multiplication, and division. For example, the + operator is used to add two numbers, and the - operator is used to subtract one number from another.
- **Assignment operators**: These operators are used to assign a value to a variable. For example, the = operator is used to assign a value to a variable, and the += operator is used to add a value to a variable and then assign the result to the variable.
- **Comparison operators**: These operators are used to compare two values and return a Boolean value (True or False) based on the result of the comparison. For example, the == operator is used to check if two values are equal, and the > operator is used to check if one value is greater than another.

In [4]:
# Perform arithmetic operations
x = 5 + 2 # 7
y = x - 3 # 4
z = y * 2 # 8

# Use assignment operators
a = 5
a += 3 # 8

# Use comparison operators
b = 5
c = 10

print(b == c) # False
print(b > c) # False


False
False


#### Hierarchy of calculations

In Python, the hierarchy of calculations refers to the order in which operations are performed in an expression. This is determined by the rules of operator precedence, which specify the order in which different types of operators are evaluated in an expression.

The general rule of operator precedence in Python is that operations enclosed in parentheses are performed first, followed by exponentiation, then multiplication and division, and finally addition and subtraction. For example, in the expression 2 + 3 * 4, the multiplication (3 * 4) is performed first, because it has a higher precedence than the addition (2 +).

In [6]:
# Use parentheses to specify the order of operations
result = (2 + 3) * 4 # 20

# Exponentiation has a higher precedence than multiplication and division
result = 2 ** 3 * 4 # 32

# Multiplication and division have the same precedence and are performed from left to right
result = 10 / 2 * 3 # 15

# Addition and subtraction have the same precedence and are performed from left to right
result = 2 + 3 - 1 # 4


Sequence of calculations for the operators is as follows:
1. Parentheses ()
2. not
3. and
4. or

#### Logical conditions

In Python, logical conditions are expressions that evaluate to either True or False. These conditions are often used in control flow statements, such as if statements, to control the flow of execution in a program.

There are several types of logical conditions in Python, including:

- **Comparison conditions**: These conditions compare two values and return True if the comparison is satisfied, or False if it is not. For example, the condition x == y checks if the value of x is equal to the value of y, and it returns True if they are equal, or False if they are not.
- **Boolean conditions**: These conditions use the Boolean values True and False directly in the condition. For example, the condition x checks if the value of x is True, and it returns True if it is, or False if it is not.
- **Membership conditions**: These conditions check if a value is a member of a sequence, such as a list or a string. For example, the condition x in y checks if the value of x is an element in the sequence y, and it returns True if it is, or False if it is not.

In [5]:
# Define two variables
x = 5
y = 10

# Comparison condition
print(x == y) # False

# Boolean condition
print(x) # True

# Membership condition
my_list = [1, 2, 3]
print(x in my_list) # False


False
5
False


#### Sequence types

Sequence types are data types that represent a group of elements that can be accessed by their position, or index, in the sequence. There are several built-in sequence types in Python, including:

 - **String**: This is the string type, and it represents a sequence of characters. For example, "Hello" is a string that contains the characters 'H', 'e', 'l', 'l', and 'o'.
 - **List**: This is the list type, and it represents a sequence of values. For example, [1, 2, 3] is a list that contains the values 1, 2, and 3.
 - **Tuple**: This is the tuple type, and it represents a sequence of values that cannot be changed. For example, (1, 2, 3) is a tuple that contains the values 1, 2, and 3.
 - **Range**: This is the range type, and it represents a sequence of integers. For example, range(5) is a range that contains the integers 0, 1, 2, 3, and 4.
  


<img src="https://i.imgur.com/Srnbhnw.png" width="200">

 - **Set**: This is the set type, and it represents a collection of unique values. For example, {1, 2, 3} is a set that contains the values 1, 2, and 3.
   - One of the main characteristics of a set is that it only allows unique items. This means that if you try to add an item to a set that already exists in the set, it will not be added again.
 - **Frozen set**: This is the frozen set type, and it represents a collection of unique values that cannot be changed. For example, frozenset({1, 2, 3}) is a frozen set that contains the values 1, 2, and 3.
 - **Dictionary**: This is the dictionary type, and it represents a collection of key-value pairs. For example, {'a': 1, 'b': 2} is a dictionary that contains the key-value pairs 'a': 1 and 'b': 2.

In [7]:
my_set = {"apple", "banana", "cherry"}
my_set.add("banana")
print(my_set)


{'banana', 'apple', 'cherry'}


##### List

List is a mutable sequence of values. It is defined using square brackets [].

In [10]:
# Объявляется через квадратные скобки и содержимое отделяется запятой
my_list = [1, 2, 3, 5]
my_list

[1, 2, 3, 5]

In [ ]:
# Можно обратиться к элементу по порядку через квадратные скобки
# Заметьте, нумерация начинается с нуля
my_list[1]  # второй элемент в списке

In [11]:
len(my_list)  # запросим длину

4

In [12]:
# Элементы списка не обязаны быть одного типа
# можно даже класть в список другой список
my_list = ['a', 1, 3.14, [1, 2]]
print(my_list)

['a', 1, 3.14, [1, 2]]


In [13]:
# К списку можно "прибавить" другой список - один список добавится к концу другого
new_list = my_list + ['h', 'e', 'l', 'l', 'o']
new_list

['a', 1, 3.14, [1, 2], 'h', 'e', 'l', 'l', 'o']

** append() **
The append() method adds an item to the end of the list.

In [14]:
# Можно добавлять через функцию append
new_list.append('g')
new_list
# Выполните несколько раз эту ячейку - каждый раз append будет добавлять в список

['a', 1, 3.14, [1, 2], 'h', 'e', 'l', 'l', 'o', 'g']

**Pop () method**

Pop () method removes the item at the given index from the list and returns the removed item. If index is not specified, pop () removes and returns the last item in the list.

In [15]:
# Можно удалять элемент, зная его индекс (0 - начало массива, 1 - второй элемент и т.д.)
# Удалит первый элемент (позиция 0)
deleted_item = new_list.pop(0)
print(deleted_item)
# pop() без аргументов удалит последний элемент
print(new_list.pop())
# Если много раз выполнить, можно опустошить весь список :)

a
g


In [16]:
new_list

[1, 3.14, [1, 2], 'h', 'e', 'l', 'l', 'o']

**Remove () method**

Remove () method removes the first item from the list whose value is equal to the given value.

In [17]:
# Можно удалять по значению, а не индексу
# Если элемента не найдется, то вылезет ошибка
print(new_list)
new_list.remove('h')
print(new_list)

[1, 3.14, [1, 2], 'h', 'e', 'l', 'l', 'o']
[1, 3.14, [1, 2], 'e', 'l', 'l', 'o']


##### **Tuple ()**

Tuple is a sequence of immutable Python objects. Tuples are sequences, just like lists. The differences between tuples and lists are, the tuples cannot be changed unlike lists and tuples use parentheses, whereas lists use square brackets.

In [18]:
# Creating a tuple
my_tuple = ("apple", "banana", "cherry")

# Accessing elements in a tuple
first_element = my_tuple[0]
last_element = my_tuple[-1]

# Iterating over a tuple
for element in my_tuple:
  print(element)

# Concatenating tuples
tuple1 = (1, 2, 3)
tuple2 = (4, 5, 6)
tuple3 = tuple1 + tuple2

# Slicing a tuple
my_tuple = ("apple", "banana", "cherry", "orange", "grape")
first_three_elements = my_tuple[:3]
last_three_elements = my_tuple[-3:]

# Checking if an element is in a tuple
my_tuple = ("apple", "banana", "cherry")
if "apple" in my_tuple:
  print("Apple is in the tuple.")


apple
banana
cherry
Apple is in the tuple.


In [19]:
# Create a tuple with three values
my_tuple = ("Alice", 22, "F")

# Create a tuple with one value
my_tuple = ("Bob",)


As you can see, to create a tuple with just one item, you have to include a comma after the item, otherwise Python will not recognize it as a tuple.

In [20]:
# Можно складывать, как и list
# при этом создастся новый (!) tuple, куда копируется сначала первый, потом второй
my_tuple + (5, 5)

('Bob', 5, 5)

Tuple does not have append(), insert(), remove(), pop() methods.

##### Set

In Python, a set is a collection of items that are unordered and unindexed. Sets are commonly used to store **only unique values** in a collection, as sets only allow for each value to be stored once. Sets are written using curly braces, with elements separated by commas. Here is an example of how to create a set in Python:

In [22]:
# create a set
my_set = {1, 2, "dog", "cat"}

# add an element to the set
my_set.add(4)

# remove an element from the set
my_set.remove(1)

# check the length of the set
len(my_set)

4

Sets are different from lists and tuples, as they are unordered and do not support indexing. This means that you cannot access items in a set using an index, and you cannot guarantee the order in which items will be stored in a set. However, sets are useful for storing unique values, and support many common operations such as union, intersection, and difference.

In Python, sets can only store immutable data types, such as numbers, strings, and tuples. This means that you cannot store lists or dictionaries in a set, as they are mutable data types. Sets also do not support duplicate values, so if you try to add a duplicate value to a set, it will not be added. Here is an example of what you can and cannot store in a set in Python:

In [26]:
# create a set
my_set = {1, 2, 3}
print(my_set)

# add a number to the set
my_set.add(4)
print(my_set)

# add a string to the set
my_set.add("hello")
print(my_set)

# add a tuple to the set
my_set.add((5, 6))
print(my_set)


{1, 2, 3}
{1, 2, 3, 4}
{1, 2, 3, 4, 'hello'}
{1, 2, 3, 4, 'hello', (5, 6)}


In [25]:
# try to add a list to the set (this will fail because lists are mutable)
my_set.add([7, 8])

# try to add a duplicate value to the set (this will fail because sets only allow unique values)
my_set.add(2)

TypeError: unhashable type: 'list'

In a `set`, **only immutable** objects can be added! This is related to its internal structure. For example, creating a set of the form `{2, "hello", []}` will not work because the third element is of type `list`, which is mutable.

Intersections, unions, and symmetric differences of sets can be taken:

- `a.intersection(b)` - is a method that returns a new set with elements common to both sets
- `a.union(b)` - is a method that returns a new set with elements from both sets
- `a.symmetric_difference(b)` - is a method that returns a new set with elements in either set, but not both

**intersection()**

In Python, the intersection of two sets can be found using the `intersection()` method. This method takes another set as an argument and returns a new set that contains only the elements that are common to both the original set and the set provided as an argument. Here is an example of how to find the intersection of two sets in Python:

In [28]:
# create two sets
set_a = {1, 2, 3}
set_b = {2, 3, 4}

# find the intersection of the two sets
intersection = set_a.intersection(set_b)

# print the intersection
print(intersection)  # prints {2, 3}


{2, 3}


**union()**

In Python, the union of two sets can be found using the `union()` method. This method takes another set as an argument and returns a new set that contains all of the elements from both the original set and the set provided as an argument. Here is an example of how to find the union of two sets in Python:

In [29]:
# create two sets
set_a = {1, 2, 3}
set_b = {2, 3, 4}

# find the union of the two sets
union = set_a.union(set_b)

# print the union
print(union)  # prints {1, 2, 3, 4}


{1, 2, 3, 4}


**symmetric_difference()**

Symmectric difference is the set of elements that are in either of the sets and not in their intersection.

In Python, the symmetric difference of two sets can be found using the `symmetric_difference()` method. This method takes another set as an argument and returns a new set that contains the elements that are present in one set but not the other. Here is an example of how to find the symmetric difference of two sets in Python:

In [30]:
# create two sets
set_a = {1, 2, 3}
set_b = {2, 3, 4}

# find the symmetric difference of the two sets
symmetric_difference = set_a.symmetric_difference(set_b)

# print the symmetric difference
print(symmetric_difference)  # prints {1, 4}


{1, 4}


`Set` does not store the index of the elements. So, you cannot access the elements of the set using the index. Sometimes `set` can mix the elements. In the latest version of Python, the elements are stored in the order of insertion but in the older versions, the elements are stored in a random order.

##### Frozen set

Frozen set is just an immutable version of a Python set object. While elements of a set can be modified at any time, elements of the frozen set remain the same after creation. Due to this, frozen sets can be used as keys in Dictionary or as elements of another set. While tuples are immutable lists, frozen sets are immutable sets.

In [39]:
# Create an empty frozen set
s = frozenset()

# Create a frozen set with some initial values
s = frozenset([1, 2, 3])

# Check if a value is in the frozen set
if 2 in s:
    print("The value 2 is in the frozen set.")

# Use the union operator to create a new frozen set with values from two other sets
s1 = frozenset([1, 2, 3])
s2 = frozenset([3, 4, 5])
s3 = s1 | s2  # s3 is a new frozen set with the values [1, 2, 3, 4, 5]

# Iterate over the values in the frozen set
for value in s:
    print(value)


The value 2 is in the frozen set.
1
2
3


In [37]:
# Try to add a new value to the frozen set (this will raise an error)
s.add(4)  # Raises an error because frozen sets are immutable

AttributeError: 'frozenset' object has no attribute 'add'

##### Dictionary

Disctionary is a collection of key-value pairs. It is mutable, and does not have a specific order. It is defined using curly braces {}.

| Имя     | Номер             |
|---------|-------------------|
| Алексей | +7 123 123-12-34  |
| Никита  | +43 321 321-32-10 |

In [31]:
# ключами может быть любой неизменяемый объект (строка, int, tuple) - как в set
name_to_number = {
    'Алексей': '+7 123 123-12-34',
    'Никита': '+43 321-32-10'
}
print(name_to_number)

{'Алексей': '+7 123 123-12-34', 'Никита': '+43 321-32-10'}


In [32]:
# По ключу можно вытащить значение
name_to_number['Алексей']

'+7 123 123-12-34'

In [33]:
# При этом, если ключа нет, то вылезет ошибка
# Можно сделать без ошибки: попросить возвращать некое значение по-умолчанию, если ключ не найден
print(name_to_number.get('Никита', 'no info'))
# Выведет +43 321-32-10, т.к. ключ "Никита" присутствует в словаре

print(name_to_number.get('Мария', 'no info'))
# Выведет "no info", т.к. ключа "Мария" нет в словаре

+43 321-32-10
no info


In [34]:
# И точно также можно редактировать через присваиваение
name_to_number['Владимир'] = 'hidden'
name_to_number['Владимир']

'hidden'

In [35]:
# Create an empty dictionary
d = {}

# Create a dictionary with some initial key-value pairs
d = {'foo': 1, 'bar': 2, 'baz': 3}

# Access the value for a specific key
value = d['foo']

# Add a new key-value pair to the dictionary
d['qux'] = 4

# Iterate over the keys and values in the dictionary
for key, value in d.items():
    print(f"{key}: {value}")

# Check if a key is in the dictionary
if 'qux' in d:
    print("The key 'qux' is in the dictionary.")


foo: 1
bar: 2
baz: 3
qux: 4
The key 'qux' is in the dictionary.


##### Nested dictionary

In Python, a nested variable is a variable that is defined inside another variable. This can be useful for organizing data and making it easier to access specific elements within a complex data structure. For example, you could have a list of dictionaries, where each dictionary contains several key-value pairs. In this case, the dictionaries would be considered nested variables, since they are defined within the list.

In [8]:
# Define a list of dictionaries
data = [
    {
        "name": "John Doe",
        "age": 32,
        "email": "johndoe@example.com"
    },
    {
        "name": "Jane Doe",
        "age": 29,
        "email": "janedoe@example.com"
    }
]

# Access the first dictionary in the list
first_dict = data[0]

# Print the contents of the dictionary
print(first_dict)


{'name': 'John Doe', 'age': 32, 'email': 'johndoe@example.com'}


To change the age in the data variable from the previous example, you can access the dictionary that you want to modify and update the value of the "age" key. Here is an example:

In [9]:
# Define a list of dictionaries
data = [
    {
        "name": "John Doe",
        "age": 32,
        "email": "johndoe@example.com"
    },
    {
        "name": "Jane Doe",
        "age": 29,
        "email": "janedoe@example.com"
    }
]

# Access the first dictionary in the list
first_dict = data[0]

# Update the age in the dictionary
first_dict["age"] = 33

# Print the contents of the dictionary to confirm the update
print(first_dict)


{'name': 'John Doe', 'age': 33, 'email': 'johndoe@example.com'}


##### Nested list
A nested list is a list that contains one or more other lists as its elements. In Python, nested lists are created by placing a list within square brackets, [], as an element of another list. Nested lists can be useful for organizing and storing data in a hierarchical structure. You can access and manipulate the elements of a nested list using the same indexing and slicing techniques that you would use with a regular list.

In [40]:
# Create an empty list
lst = []

# Create a list with some initial values
lst = [1, 2, 3]

# Access the first element in the list
first_element = lst[0]

# Add a new element to the end of the list
lst.append(4)

# Create a nested list
nested_lst = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

# Access an element in the nested list
element = nested_lst[1][2]  # This is the element 6

# Add a new nested list to the outer list
nested_lst.append([10, 11, 12])

# Iterate over the elements in the nested list
for sublist in nested_lst:
    for element in sublist:
        print(element)


1
2
3
4
5
6
7
8
9
10
11
12


### Conditional statements and loops

https://www.youtube.com/watch?v=ucsLI2ITH64&ab_channel=karpov.courses

<iframe width="560" height="315" src="https://www.youtube.com/embed/ucsLI2ITH64" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

#### For & While loops

In Python, `for` loops are used to iterate over a sequence of elements, such as a list or a string. A `for` loop begins with the keyword `for`, followed by a variable name, followed by the keyword `in`, and then the name of the sequence that you want to iterate over. The body of the for loop is indented, and typically includes one or more statements that operate on the elements of the sequence. Here is an example of a for loop in Python that iterates over a list of numbers and prints each number to the screen:

In [1]:
numbers = [1, 2, 3, 4, 5]

for number in numbers:
    print(number)


1
2
3
4
5


This for loop will print the numbers 1, 2, 3, 4, and 5 to the screen, each on a separate line.

On the other hand, a `while` loop in Python is used to repeat a block of code as long as a certain condition is met. A `while` loop begins with the keyword `while`, followed by a condition, and then a colon. The body of the `while` loop is indented and contains one or more statements that are executed repeatedly as long as the condition remains true. Here is an example of a `while` loop in Python that counts down from 10 to 1 and prints each number to the screen:

In [5]:
number = 5

while number > 0:
    print(number)
    number = number - 1


5
4
3
2
1


In general, you would use a `for` loop when you know in advance how many times you want to iterate over a sequence, such as when you are looping over a list of numbers or a string. You would use a `while` loop when you want to repeat a block of code as long as a certain condition remains true, such as when you are waiting for user input or when you are performing an operation that may take an unknown amount of time to complete.

It is generally not correct to say that one type of loop is "faster" than the other, since the performance of a loop in Python depends on various factors, such as the specific operations being performed inside the loop and the size and type of the sequence being iterated over. In general, a `for` loop may be slightly faster than a `while` loop when iterating over a large sequence, since the `for` loop can access the elements of the sequence directly, whereas the `while` loop has to use a separate variable to keep track of the current position in the sequence. However, the difference in performance is typically very small and may not be noticeable in most cases. It is more important to choose the type of loop that is most appropriate for the task at hand, based on whether you know in advance how many times you want to iterate over a sequence or whether you want to repeat a block of code as long as a certain condition remains true.

<img src="https://i.imgur.com/J1HMXmb.png" width="300"><img src="https://i.imgur.com/KGH89El.png" width="240"><img src="https://media.tenor.com/kEdtBWkn2vEAAAAC/the-rock-look.gif" width="100">

In [3]:
print('Генерируем четные числа')
# range(start, stop, step) генерирует от start до (stop - 1) с шагом step
even_numbers = []
for i in range(0, 14, 2):
    even_numbers.append(i)
print(even_numbers)

Генерируем четные числа
[0, 2, 4, 6, 8, 10, 12]


Here is an example of how the code can be rewritten using a while loop:

In [7]:
print('Генерируем четные числа')

even_numbers = []
i = 0
while i < 14:
    even_numbers.append(i)
    i += 2
print(even_numbers)


Генерируем четные числа
[0, 2, 4, 6, 8, 10, 12]


In [4]:
# Представим, что нас попросили подсчитать, сколько раз число делится на 2
number = 34624

# Мы заранее не можем знать, сколько раз число поделится на 2 - тут и выручает while
while number % 2 == 0:  # кстати, number % 2 == 0 - имеет булевый тип
    print('делится на 2, делим')
    number //= 2 # то же самое, что number = number // 2
print(number)

делится на 2, делим
делится на 2, делим
делится на 2, делим
делится на 2, делим
делится на 2, делим
делится на 2, делим
541


Here is an example of how the code can be rewritten using a for loop:

In [6]:
number = 34624

for i in range(number):
    if number % 2 == 0:
        print('делится на 2, делим')
        number //= 2
    else:
        break
print(number)


делится на 2, делим
делится на 2, делим
делится на 2, делим
делится на 2, делим
делится на 2, делим
делится на 2, делим
541


#### if statement

In Python, the `if` statement is not the only control flow statement that allows you to execute code based on a certain condition. There are several other statements that you can use to add more conditions to your code and control the flow of execution more precisely.

The `if` statement is used to execute a block of code only if a certain condition is satisfied. Here is an example of an `if` statement in Python:

In [8]:
if x > 10:
    # execute this code
    print("x is greater than 10")


NameError: name 'x' is not defined

Before using a variable in `if` statement, we need to declare a variable.

In [9]:
x = 69

if x > 10:
    # execute this code
    print("x is greater than 10")

x is greater than 10


The if statement is often used in combination with the else statement, which allows you to execute a different block of code if the condition in the if statement is not satisfied. Here is an example of an if statement with an else clause:

In [10]:
if x > 10:
    # execute this code if x > 10
    print("x is greater than 10")
else:
    # execute this code if x <= 10
    print("x is less than or equal to 10")


x is greater than 10


In this example, the code inside the `if` statement will be executed if `x` is greater than 10, and the code inside the `else` clause will be executed if `x` is less than or equal to 10.

The if statement can also be used in combination with the elif (short for "`else if`") clause, which allows you to specify additional conditions to check. Here is an example of an `if` statement with an `elif` clause:

In [11]:
if x > 10:
    # execute this code if x > 10
    print("x is greater than 10")
elif x > 5:
    # execute this code if x > 5 but x <= 10
    print("x is greater than 5 but less than or equal to 10")
else:
    # execute this code if x <= 5
    print("x is less than or equal to 5")


x is greater than 10


You can use as many `elif` clauses as you want in an if statement. The code inside each `elif` clause will be checked in the order in which they appear, and the first one that is satisfied will be executed. If none of the `elif` clauses are satisfied, the code inside the `else` clause (if there is one) will be executed.

Overall, the if statement is a powerful tool that allows you to control the flow of execution in your Python code based on certain conditions, and the else and elif clauses provide additional flexibility for specifying more complex conditions.

In [12]:
# А теперь поищем счета с отрицательным балансом
accounts_balance = [1482.0, 28182.12, -124.42, 85.3, -23.5, 82]  # входные данные

# сюда запишем ответ
negative_accounts = []
# проходимся по всем элементам во входных данных
for balance in accounts_balance:
    # если значение текущего элемента < 0, то идем во внутрь `if`
    if balance < 0:
        print(f'{balance} отрицателен, записываем в ответ')
        negative_accounts.append(balance)
    # выполнится в любом случае, т.к. не находится под властью `if` (отступы важны!)
    print(f'закончили обрабатывать баланс {balance}')
    
print(f'Отрицательные балансы: {negative_accounts}')

закончили обрабатывать баланс 1482.0
закончили обрабатывать баланс 28182.12
-124.42 отрицателен, записываем в ответ
закончили обрабатывать баланс -124.42
закончили обрабатывать баланс 85.3
-23.5 отрицателен, записываем в ответ
закончили обрабатывать баланс -23.5
закончили обрабатывать баланс 82
Отрицательные балансы: [-124.42, -23.5]


`input()` is a built-in function in Python that allows a user to enter input from the keyboard. This input can then be stored in a variable and used in your program.

In [14]:
# ask the user to enter their name
name = input("Please enter your name: ")

# print a greeting to the user
print("Hello, " + name)


Hello, asdf


It is important to note that the `input()` function always returns a string, even if the user enters a number. This means that if you want to use the input as a number (integer or floating-point), you will need to convert it to the appropriate data type using one of the built-in functions such as `int()` or `float()`. Here is an example:

In [15]:
# ask the user to enter a number
number = input("Please enter a number: ")

# convert the input to an integer and print it
number = int(number)
print(number)


123


Overall, the `input()` function is a useful tool for allowing a user to enter input in your Python program, and it can be combined with other built-in functions to convert the input to the appropriate data type for your needs.

In [13]:
# input() - это запрос ввода от пользователя
# У вас появится внизу строка, куда можно ввести что угодно и это запишется в `name`
# Попробуйте ввести в строку "Москва" и "Лондон" без кавычек (запустите ячейку два раза)
name = input()
if name == 'Москва':
    print('Столица России')
# else if
elif name == 'Лондон':
    print('Столица Великобритании')
else:
    print('Неизвестно')

Неизвестно


#### Break/continue statement

In Python, the `break` and `continue` statements are used to control the flow of execution in loops (`for` and `while` loops).

The `break` statement is used to exit a loop immediately, without executing any of the remaining code in the loop. For example, consider the following code that uses a `while` loop to print the numbers from 1 to 10:

In [17]:
# initialize the variable i to 1
i = 1

# start a while loop that will run until i > 10
while i < 10:
    # print the current value of i
    print(i)

    # increment i by 1
    i += 1


1
2
3
4
5
6
7
8
9


In [19]:
# initialize the variable i to 1
i = 1

# start a while loop that will run until i > 10
while i < 10:
    # check if i is equal to 5
    if i == 5:
        # if i is equal to 5, stop the loop and exit immediately
        break

    # print the current value of i
    print(i)

    # increment i by 1
    i += 1


1
2
3
4


In this code, the `break` statement is used to stop the `while` loop and exit immediately when the value of i is equal to 5. This means that the loop will only run four times (1, 2, 3, 4) and the value of i will not be printed when it reaches 5.

The `continue` statement is similar to the `break` statement, but it is used to skip the remaining code in the current iteration of the loop and move on to the next iteration. For example, consider the following code that uses a for loop to print the numbers from 1 to 10:

In [20]:
# use a for loop to iterate over the numbers from 1 to 10
for i in range(1, 11):
    # print the current value of i
    print(i)


1
2
3
4
5
6
7
8
9
10


In this code, the `for` loop will iterate over the numbers from 1 to 10 and print each number on the screen. However, if you want to skip certain numbers (for example, if you only want to print the even numbers), you can use the `continue` statement. Here is an example:

In [21]:
# use a for loop to iterate over the numbers from 1 to 10
for i in range(1, 11):
    # check if i is odd (not divisible by 2)
    if i % 2 != 0:
        # if i is odd, skip the remaining code in this iteration and move on to the next iteration
        continue

    # print the current value of i (only reached if i is even)
    print(i)


2
4
6
8
10


In this code, the `continue` statement is used to skip the remaining code in the current iteration of the `for` loop if `i` is odd (not divisible by 2). This means that only the even numbers (2, 4, 6, 8, 10)

In [22]:
for i in range(5):
    for j in range(7):
        print(i, j)
        if j > 2:
            # Выкинет только из внутреннего цикла
            # Увидим, что i в выводе принимает значения 3, 4
            # это ожидаемо: break происходит после print(i, j) и одна итерация "просачивается"
            break
    print('увеличиваем i')

0 0
0 1
0 2
0 3
увеличиваем i
1 0
1 1
1 2
1 3
увеличиваем i
2 0
2 1
2 2
2 3
увеличиваем i
3 0
3 1
3 2
3 3
увеличиваем i
4 0
4 1
4 2
4 3
увеличиваем i


`Continue` statement will throw one iteration and will not execute the remaining code in the current iteration of the loop and move on to the next iteration.

In [23]:
for feature in ['make sandwitch', 'make coffee', 'watch TV', 'wash plate']:
    if feature == 'make coffee':
        # Пропустим одну итерацию в цикле
        print('Coffee machine is broken, skipping :(')
        continue
    # else уже не нужен - если условие не истинно
    # то все равно выйдем на print ниже
    # а если истинно - continue гарантирует, что дальше цикл выполняться не будет
    print(f'starting action: {feature}')

starting action: make sandwitch
Coffee machine is broken, skipping :(
starting action: watch TV
starting action: wash plate


For what might this be needed? Sometimes in the logic of the loop there are deviations: for example, a certain combination of parameters cannot be processed by the loop - then we use `continue`.

`break` is most often used in `while` loops as a specification of when to stop the loop.

Example: in machine learning there is an algorithm of gradient descent that "teaches" the model on data. The "training" process goes in steps, and often the process needs to be interrupted if the new step did not lead to an improvement in the model - in such cases you can use `break` to exit.

### Lesson 1: homework

#### Базовый синтаксис

Создайте три переменных name, age и is_student и впишите в них любые имя, возраст и True/False соответственно.

Затем распечатайте эти переменные, передав их тремя аргументами в функцию print, либо вызвав print три раза.

Создайте комментарий в любом месте программы и напишите туда дату решения задачи. Это ни на что не повлияет — просто попрактикуемся в написании комментариев.

In [24]:
# December 15, 2022

name = "Alexey"
is_student = True
age = 30

print(name, is_student, age)


Alexey True 30


#### Арифметика: поход в магазин
В следующих степах мы попробуем использовать переменные для простой арифметики.

Представим, что у нас есть `total_money` рублей и мы идем в магазин покупать варенье про запас. Каждая банка варенья стоит `price` рублей (при этом `price` > 0). Наша задача — подсчитать, сколько банок мы сможем купить.

NB! Переменные `total_money` и `price` уже заданы в ЛМС, их не нужно задавать самостоятельно. Ваша задача — подсчитать ответ на задачу, используя эти готовые переменные.

Запишите результат в переменную jar_count.

In [26]:
total_money = 420
price = 69

In [28]:
# December 15, 2022

jar_count = total_money // price

print(jar_count)

6


#### Арифметика: квадратные уравнения

<div class="viewer--28ZF"><h2>&gt; Арифметика:&nbsp;квадратные уравнения</h2><p class="p--1qUA">Продолжаем знакомство с арифметикой в Python. Теперь мы научимся работать с дробными числами, возводить в степень и расставлять приоритет операций с помощью скобок.</p><hr><p class="p--1qUA">В практических приложениях математики часто встречаются <i>квадратные уравнения </i>— это уравнения вида:</p><p><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8974em;vertical-align:-0.0833em;"></span><span class="mord mathnormal">a</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.7778em;vertical-align:-0.0833em;"></span><span class="mord mathnormal">b</span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.4306em;"></span><span class="mord mathnormal">c</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">0</span></span></span></span></p><p class="p--1qUA">Здесь <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">a</span></span></span></span></code>, <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6944em;"></span><span class="mord mathnormal">b</span></span></span></span></code> и <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">c</span></span></span></span></code> — некие числа. Вот пример такого уравнения:</p><p><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8974em;vertical-align:-0.0833em;"></span><span class="mord">3</span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.7278em;vertical-align:-0.0833em;"></span><span class="mord">7</span><span class="mord mathnormal">x</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">10</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.6444em;"></span><span class="mord">0</span></span></span></span></p><p class="p--1qUA">В примере <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">a</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">3</span></span></span></span></code>, <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6944em;"></span><span class="mord mathnormal">b</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">7</span></span></span></span></code>, <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">c</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.7278em; vertical-align: -0.0833em;"></span><span class="mord">−</span><span class="mord">10</span></span></span></span></code>.</p><p class="p--1qUA">У таких уравнений может быть два, одно или ноль решений. Мы для простоты будем рассматривать только те уравнения, которые имеют два решения. Если решений два, то они задаются формулами:</p><p><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.5806em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3011em;"><span style="top:-2.55em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:2.3394em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.6534em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span><span class="mord mathnormal">a</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.74em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">−</span><span class="mord mathnormal">b</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.9134em;"><span class="svg-align" style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="mord" style="padding-left:0.833em;"><span class="mord"><span class="mord mathnormal">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.7401em;"><span style="top:-2.989em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord">4</span><span class="mord mathnormal">a</span><span class="mord mathnormal">c</span></span></span><span style="top:-2.8734em;"><span class="pstrut" style="height:3em;"></span><span class="hide-tail" style="min-width:0.853em;height:1.08em;"><svg xmlns="http://www.w3.org/2000/svg" width="400em" height="1.08em" viewBox="0 0 400000 1080" preserveAspectRatio="xMinYMin slice"><path d="M95,702
c-2.7,0,-7.17,-2.7,-13.5,-8c-5.8,-5.3,-9.5,-10,-9.5,-14
c0,-2,0.3,-3.3,1,-4c1.3,-2.7,23.83,-20.7,67.5,-54
c44.2,-33.3,65.8,-50.3,66.5,-51c1.3,-1.3,3,-2,5,-2c4.7,0,8.7,3.3,12,10
s173,378,173,378c0.7,0,35.3,-71,104,-213c68.7,-142,137.5,-285,206.5,-429
c69,-144,104.5,-217.7,106.5,-221
l0 -0
c5.3,-9.3,12,-14,20,-14
H400000v40H845.2724
s-225.272,467,-225.272,467s-235,486,-235,486c-2.7,4.7,-9,7,-19,7
c-6,0,-10,-1,-12,-3s-194,-422,-194,-422s-65,47,-65,47z
M834 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.1266em;"><span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span></span></span></span></span></span></p><p><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.5806em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3011em;"><span style="top:-2.55em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:2.3394em;vertical-align:-0.686em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.6534em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">2</span><span class="mord mathnormal">a</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.74em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">−</span><span class="mord mathnormal">b</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.9134em;"><span class="svg-align" style="top:-3em;"><span class="pstrut" style="height:3em;"></span><span class="mord" style="padding-left:0.833em;"><span class="mord"><span class="mord mathnormal">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.7401em;"><span style="top:-2.989em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord">4</span><span class="mord mathnormal">a</span><span class="mord mathnormal">c</span></span></span><span style="top:-2.8734em;"><span class="pstrut" style="height:3em;"></span><span class="hide-tail" style="min-width:0.853em;height:1.08em;"><svg xmlns="http://www.w3.org/2000/svg" width="400em" height="1.08em" viewBox="0 0 400000 1080" preserveAspectRatio="xMinYMin slice"><path d="M95,702
c-2.7,0,-7.17,-2.7,-13.5,-8c-5.8,-5.3,-9.5,-10,-9.5,-14
c0,-2,0.3,-3.3,1,-4c1.3,-2.7,23.83,-20.7,67.5,-54
c44.2,-33.3,65.8,-50.3,66.5,-51c1.3,-1.3,3,-2,5,-2c4.7,0,8.7,3.3,12,10
s173,378,173,378c0.7,0,35.3,-71,104,-213c68.7,-142,137.5,-285,206.5,-429
c69,-144,104.5,-217.7,106.5,-221
l0 -0
c5.3,-9.3,12,-14,20,-14
H400000v40H845.2724
s-225.272,467,-225.272,467s-235,486,-235,486c-2.7,4.7,-9,7,-19,7
c-6,0,-10,-1,-12,-3s-194,-422,-194,-422s-65,47,-65,47z
M834 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.1266em;"><span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.686em;"><span></span></span></span></span></span><span></span></span></span></span></span></p><p class="p--1qUA">Выглядит пугающе! Но на деле две формулы отличаются только заменой знака "<code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6667em; vertical-align: -0.0833em;"></span><span class="mord">+</span></span></span></span></code>" на знак "<code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.6667em; vertical-align: -0.0833em;"></span><span class="mord">−</span></span></span></span></code>" перед квадратным корнем. Если вы забыли эти формулы, ничего страшного — сейчас нам они нужны только для практики в арифметике :)</p><hr><p class="p--1qUA"><b>Наша задача</b>: по известным <code class="klms-inline-code" data-lang="common">a</code>, <code class="klms-inline-code" data-lang="common">b</code>, <code class="klms-inline-code" data-lang="common">c</code> подсчитать оба корня этого уравнения. Как и в прошлом задании, переменные <code class="klms-inline-code" data-lang="common">a</code>, <code class="klms-inline-code" data-lang="common">b</code> и <code class="klms-inline-code" data-lang="common">c</code> уже заданы в ЛМС, их не нужно задавать самостоятельно.</p><p class="p--1qUA">Сохраните оба решения уравнения в переменные с названиями <code class="klms-inline-code" data-lang="common">x_1</code> и <code class="klms-inline-code" data-lang="common">x_2</code> соответственно. Порядок не важен — если уравнение имеет решения 2 и 5, то оба варианта <code class="klms-inline-code" data-lang="common">x_1 = 2,</code> <code class="klms-inline-code" data-lang="common">x_2 = 5</code> и <code class="klms-inline-code" data-lang="common">x_1 = 5</code>, <code class="klms-inline-code" data-lang="common">x_2 = 2</code> будут зачтены.</p><p class="p--1qUA">Для взятия квадратного корня можно возводить в степень <code class="klms-inline-code" data-lang="common">0.5</code>. Скажем, если мы хотим подсчитать <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 1.04em; vertical-align: -0.1328em;"></span><span class="mord sqrt"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.9072em;"><span class="svg-align" style="top: -3em;"><span class="pstrut" style="height: 3em;"></span><span class="mord" style="padding-left: 0.833em;"><span class="mord">9</span></span></span><span class="" style="top: -2.8672em;"><span class="pstrut" style="height: 3em;"></span><span class="hide-tail" style="min-width: 0.853em; height: 1.08em;"><svg width="400em" height="1.08em" viewBox="0 0 400000 1080" preserveAspectRatio="xMinYMin slice"><path d="M95,702
c-2.7,0,-7.17,-2.7,-13.5,-8c-5.8,-5.3,-9.5,-10,-9.5,-14
c0,-2,0.3,-3.3,1,-4c1.3,-2.7,23.83,-20.7,67.5,-54
c44.2,-33.3,65.8,-50.3,66.5,-51c1.3,-1.3,3,-2,5,-2c4.7,0,8.7,3.3,12,10
s173,378,173,378c0.7,0,35.3,-71,104,-213c68.7,-142,137.5,-285,206.5,-429
c69,-144,104.5,-217.7,106.5,-221
l0 -0
c5.3,-9.3,12,-14,20,-14
H400000v40H845.2724
s-225.272,467,-225.272,467s-235,486,-235,486c-2.7,4.7,-9,7,-19,7
c-6,0,-10,-1,-12,-3s-194,-422,-194,-422s-65,47,-65,47z
M834 80h400000v40h-400000z"></path></svg></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height: 0.1328em;"><span class=""></span></span></span></span></span></span></span></span></code>, то можем воспользоваться <code class="klms-inline-code" data-lang="common">9**0.5</code> — это выражение вернет <code class="klms-inline-code" data-lang="common">3.0</code> (Python считает, что взятие квадратного корня всегда возвращает дробное число).</p><p class="p--1qUA">На всякий случай напомним порядок операций при вычислении <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.5806em; vertical-align: -0.15em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span class="" style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">1</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span class=""></span></span></span></span></span></span></span></span></span></code> и <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.5806em; vertical-align: -0.15em;"></span><span class="mord"><span class="mord mathnormal">x</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height: 0.3011em;"><span class="" style="top: -2.55em; margin-left: 0em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height: 0.15em;"><span class=""></span></span></span></span></span></span></span></span></span></code>:</p><ol><li class="p--1qUA">Взять <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8141em;"></span><span class="mord"><span class="mord mathnormal">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.8141em;"><span class="" style="top: -3.063em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span></code></li><li class="p--1qUA">Подсчитать <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8974em; vertical-align: -0.0833em;"></span><span class="mord"><span class="mord mathnormal">b</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height: 0.8141em;"><span class="" style="top: -3.063em; margin-right: 0.05em;"><span class="pstrut" style="height: 2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">4</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">∗</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.4653em;"></span><span class="mord mathnormal">a</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">∗</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">c</span></span></span></span></code></li><li class="p--1qUA">Взять корень из результата в п.2</li><li class="p--1qUA">Взять <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.7778em; vertical-align: -0.0833em;"></span><span class="mord">−</span><span class="mord mathnormal">b</span><span class="mspace" style="margin-right: 0.2222em;"></span><span class="mbin">−</span><span class="mspace" style="margin-right: 0.2222em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord cyrillic_fallback">п</span><span class="mord">.3</span></span></span></span></code></li><li class="p--1qUA">Поделить результат на удвоенное <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.4306em;"></span><span class="mord mathnormal">a</span></span></span></span></code></li></ol><p class="p--1qUA">В проверяющей системе числа <code class="klms-inline-code" data-lang="common">a</code>, <code class="klms-inline-code" data-lang="common">b</code> и <code class="klms-inline-code" data-lang="common">c</code> будут подобраны так, чтобы уравнение гарантированно имело два решения и <code class="view-katex-processed" data-lang="latex"><span class="katex"><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height: 0.8889em; vertical-align: -0.1944em;"></span><span class="mord mathnormal">a</span><span class="mspace" style="margin-right: 0.2778em;"></span><span class="mrel"><span class="mrel"><span class="mord vbox"><span class="thinbox"><span class="rlap"><span class="strut" style="height: 0.8889em; vertical-align: -0.1944em;"></span><span class="inner"><span class="mord"><span class="mrel"></span></span></span><span class="fix"></span></span></span></span></span><span class="mrel">=</span></span><span class="mspace" style="margin-right: 0.2778em;"></span></span><span class="base"><span class="strut" style="height: 0.6444em;"></span><span class="mord">0</span></span></span></span></code>.</p><p class="p--1qUA">Если нужна помощь,&nbsp;напишите <a href="https://discord.com/channels/1049642177002745926/1050148717132529777/1050148720404086794" class="view-links-processed" target="_blank">сюда</a>.</p><p class="p--1qUA"><a href="https://abrupt-moose-9e7.notion.site/1-4-PYTHON-f56aa696c8444ad88288883bc293d9c5" class="view-links-processed" target="_blank">Подсказки</a><font color="#ff5533"></font></p><div class="pb-3 pt-3"><iframe loading="eager" width="100%" class="iframe--2exJ" src="https://lab.karpov.courses/kchecker/iframe/?data=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbnRpdHlfdHlwZSI6InN0ZXAiLCJlbnRpdHlfaWQiOjI0OTAyNywib3BlcmF0aW9uIjoidmlldyIsInNldHRpbmdzIjp7ImljIjoiY29kZS1zbGFzaCIsInR5cGUiOiJjb2RlIiwiYXV0b19wYXNzIjpmYWxzZX0sImlkIjoiOTIxMSIsImNsaWVudF9pZCI6IkQzSEV1aTZNSGZBbU9vb2lrQ0laUmZRWmZyQ2hXUEFac0lYSTlLc1UiLCJhY2Nlc3NfdG9rZW4iOiJ1V1c3QlNnSWNHbjN4TGw2ckdVa3gySmVqVWE0Z3IiLCJleHAiOjE2NzExNjk5OTEsInVzZXJfaWQiOjE2MzMxfQ.yXII2IkdglGN3kpKXXxcnOoNI2CjCpJ3LHPLjJQDyGU&amp;nonce=c5e7064bf8a34e91965d0ee52f78fb90" height="349.953125px"></iframe></div><hr><p>Максимальное количество баллов - 1 за решение после 26 декабря 2022 г., 20:00</p></div>

In [29]:
# December 15, 2022

# a*x**2 + b*x + c = 0

a = 3
b = 7
c = -10

In [31]:
x_1 = (-b + (b**2 - 4*a*c)**0.5) / (2*a)
x_2 = (-b - (b**2 - 4*a*c)**0.5) / (2*a)

print(x_1, x_2)

1.0 -3.3333333333333335


#### Logical Operators

What will python return on `True and False or True`?

  * **True**
  * False
  * None
  * Error

In the expression `true` and `false` or `true`, the and operator has higher precedence than the or operator, so it will be evaluated first. The expression `true` and `false` evaluates to `False`, so the entire expression becomes `False` or `true`. The or operator then evaluates the second operand, which is `True`, and returns `True` as the result of the expression.

Therefore, the expression `true` and `false` or `true` would evaluate to `True` in Python.

#### Коллекции
Давайте теперь посмотрим, как Python работает с нечисленными данными.

Допустим, у нас задача сделать книжку для записи дел на сегодня. Нам хочется, чтобы порядок дел в ней сохранялся (то есть в каком порядке дела добавили, в таком они и будут записаны в книжке). Также мы хотим уметь удалять элементы из этой книжки. 

Как думаете, какой тип данных подойдет для этого?

    * **Список**
    * Словарь
    * Множество
    * Строка

#### Коллекции: объединение списков
Все готово к тому, чтобы вести дела. У нас есть два списка `tasks_my` и `tasks_friend`.

Теперь необходимо объединить эти списки в один (любым способом, сначала `tasks_my`, потом `tasks_friend`) и сохраните результат в `tasks_all`.

NB! Переменные `tasks_my` и `tasks_friend` уже загружены в ЛМС, создавать их не нужно.

In [32]:
tasks_my = ['do nothing', 'go to sleep', 'go to the gym']
tasks_friend = ['go watch a movie', 'go to the gym', 'go to the store']

In [33]:
# Combine two lists
tasks_all = tasks_my + tasks_friend

print(tasks_all)

['do nothing', 'go to sleep', 'go to the gym', 'go watch a movie', 'go to the gym', 'go to the store']


#### Коллекции: система приоритетов
Список дел у нас готов. Давайте теперь добавим ему систему приоритетов.

Чтобы задать приоритеты, нужно хранить не только задачу, но и число (скажем, 0, 1 или 5) . Чем ниже число, тем приоритетнее задача. Вообще, для такого списка можно использовать разные структуры данных, но мы поставим требование: необходимо иметь удобный способ вытащить все задачи с определенным приоритетом, и вытащить их в порядке добавления.

Как думаете, какая структура данных для этого подойдет?

- Задать множество {"задача 1", "задача 5"}
- Задать dict, где ключом будет приоритет, а значением — одна задача
- **Задать dict, где ключом будет приоритет, а значением — список дел на этом приоритете**
- Задать dict, где ключом будет задача, а значением — ее приоритет
- Задать список из пар (кортежей длины 2) такого вида: [(0, "задача 1"), (5, "задача с приоритетом 5")]

#### Коллекции: записная книжка

<div class="viewer--28ZF"><h2>&gt; Коллекции: записная книжка</h2><p class="p--1qUA">Теперь, когда мы определились со структурой данных, мы можем создать свою записную книжку.</p><p class="p--1qUA">Ее содержание должно быть следующим:&nbsp;</p><table class="table table-bordered"><tbody><tr><td><b>Приоритет</b></td><td><b>Задача</b></td></tr><tr><td>1</td><td>Полить цветы</td></tr><tr><td>0</td><td>Покормить кота</td></tr><tr><td>1</td><td>Забрать посылку</td></tr><tr><td>2</td><td>Почитать книгу по программированию</td></tr><tr><td>3</td><td>Ответить на письмо двоюродной тети</td></tr></tbody></table><p class="p--1qUA"><b>Задача: </b>Напишите код, создающий словарь, и сохраните его в переменную&nbsp;<code class="klms-inline-code" data-lang="common">tasks</code>.</p><p class="p--1qUA"><font color="#ff5533">NB!</font> Учтите, что задачи одинакового приоритета должны быть в том же порядке, что и в таблице!</p><p class="p--1qUA"><font color="#ff5533">NB!</font> Ключи в этом словаре должны иметь целочисленный тип.</p><hr><p class="p--1qUA">Это может быть немного трудно, поэтому для решения задачи можно повторить:</p><ul><li class="p--1qUA"><a href="/learning/179/module/1941/lesson/18325/52807/249017/" class="view-links-processed">Как создавать список</a></li><li class="p--1qUA"><a href="/learning/179/module/1941/lesson/18325/52807/249019/" class="view-links-processed">Как создавать ключ в словаре</a></li></ul><p class="p--1qUA">Выполняя код в юпитер ноутбуке, вы можете выводить промежуточный результат через <code class="klms-inline-code" data-lang="common">print(tasks)</code> — это удобный способ следить, на правильном ли вы пути.</p><p class="p--1qUA">Если нужна помощь,&nbsp;напишите <a href="https://discord.com/channels/1049642177002745926/1050148759901847612/1050148763232124999" class="view-links-processed" target="_blank">сюда</a>.</p><p class="p--1qUA"><a href="https://abrupt-moose-9e7.notion.site/1-10-PYTHON-c26f644902714b44831347d2fa04bb68" class="view-links-processed" target="_blank">Подсказки</a></p><div class="pb-3 pt-3"><iframe loading="eager" width="100%" class="iframe--2exJ" src="https://lab.karpov.courses/kchecker/iframe/?data=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbnRpdHlfdHlwZSI6InN0ZXAiLCJlbnRpdHlfaWQiOjI0OTAzMywib3BlcmF0aW9uIjoidmlldyIsInNldHRpbmdzIjp7ImljIjoiY29kZS1zbGFzaCIsInR5cGUiOiJjb2RlIiwiYXV0b19wYXNzIjpmYWxzZX0sImlkIjoiOTIxNSIsImNsaWVudF9pZCI6IkQzSEV1aTZNSGZBbU9vb2lrQ0laUmZRWmZyQ2hXUEFac0lYSTlLc1UiLCJhY2Nlc3NfdG9rZW4iOiJ1V1c3QlNnSWNHbjN4TGw2ckdVa3gySmVqVWE0Z3IiLCJleHAiOjE2NzExNzM4NzUsInVzZXJfaWQiOjE2MzMxfQ.O77AVhdowgTeFhZ1TrNIVeMsk02DtByctKisD6UfYEw&amp;nonce=a09a02b99fb041cfa279f142f45b1f9e" height="349.953125px"></iframe></div><hr><p>Максимальное количество баллов - 1.5 за решение после 26 декабря 2022 г., 20:00</p></div>

In [36]:
# make a dictionary for tasks and priorities

tasks = {
    0: ['Покормить кота'],
    1: ['Полить цветы', 'Забрать посылку'], 
    2: ['Почитать книгу по программированию'], 
    3: ['Ответить на письмо двоюродной тети']
    }

print(tasks)

{0: ['Покормить кота'], 1: ['Полить цветы', 'Забрать посылку'], 2: ['Почитать книгу по программированию'], 3: ['Ответить на письмо двоюродной тети']}


#### Условные переходы: основы
Отлично! Теперь у нас есть словарь tasks, в нем содержатся задачи разного приоритета. 

Сейчас мы хотим выполнить простую проверку: если есть задачи нулевого приоритета, программа должна выводить строку "есть срочные дела", иначе —  "можно отдохнуть".

In [38]:
# Check if we have values in the dictionary with kjeys 0
if 0 in tasks:
    print('есть срочные дела')
else:
    print('можно отдохнуть')


есть срочные дела


#### Циклы: основы
В этом задании мы продолжим работать со словарем tasks, в котором содержатся задачи разного приоритета, и напишем свой первый цикл!

Проитерируйтесь по ключам словаря и добавьте его значения (списки задач) в список values.

NB! Переменная tasks уже загружена в ЛМС, создавать ее не нужно.

In [45]:
values = []
# For loop for dictionary and store it in a list
for key, value in tasks.items():
    values.append(value)
    
print(values)

[['Покормить кота'], ['Полить цветы', 'Забрать посылку'], ['Почитать книгу по программированию'], ['Ответить на письмо двоюродной тети']]


#### Циклы: основы
Немного усложним задачу. Прежде мы итерировались по ключам словаря и сохраняли его значения —  списки задач. Теперь мы хотим сохранить не списки задач, а **сами задачи**. Для этого необходимо проитерироваться по всем объектам в значениях словаря и добавить задачи в список `doings`.

NB! Переменная `tasks` уже загружена в ЛМС, создавать ее не нужно.

In [46]:
doings = []

# For loop for dictionary and store values in a list but with strings instead of lists
for key, value in tasks.items():
    for item in value:
        doings.append(item)
        
print(doings)

['Покормить кота', 'Полить цветы', 'Забрать посылку', 'Почитать книгу по программированию', 'Ответить на письмо двоюродной тети']


#### Все вместе: ищем кота в своих делах
Это задание включает часть идей реализации из предыдущего, но теперь, помимо итерации по всем ключам и значениям, вам необходимо отобрать только те задачи, в которых есть подстрока "кот", и добавить их в список `answer`.

NB! Переменная `tasks` уже загружена в ЛМС, создавать ее не нужно.

In [47]:
answer = []

# find all variables in the dictionary that contain the word "кот" and store them in a list answer
for key, value in tasks.items():
    for item in value:
        if 'кот' in item:
            answer.append(item)

print(answer)
    

['Покормить кота']


#### Continue и break: спасаемся от работы
Следующее задание похоже на предыдущее. Мы будем итерироваться по ключам словаря `tasks`, брать задания оттуда и по одному добавлять их в список `answer`. Вот только мы хотим уберечь себя от большого количества задач, поэтому в процессе будем проверять, не взяли ли мы уже на себя слишком много работы.

Сегодня мы ленимся, поэтому готовы взять только самые срочные задания. Конкретно — только те, приоритет которых меньше двух.

Если длина списка `answer` достигает двух, то мы решаем, что дел на сегодня уже достаточно, и прерываем выполнение цикла. И еще важный момент — в список нужно добавлять только задачи в которых содержится подстрока `"кот"`. 

В конце должна быть заполненная переменная `answer`, в которой лежат задачи (не списки задач), полученные по алгоритму выше. Печатать список не обязательно - вы наверняка его уже много раз распечатали, пока разрабатывали код :)

NB! Переменная `tasks` уже загружена в ЛМС, создавать ее не нужно.

In [64]:
tasks = {
    0: ['Покормить кота'],
    1: ['Полить цветы', 'Забрать посылку'], 
    2: ['Почитать книгу по программированию', 'Покормить кота 2'], 
    3: ['Ответить на письмо двоюродной тети']
    }

In [66]:
answer = []

for key, value in tasks.items():
    # check if the proirity is < 2 and check if it has a value "кот"
    if key <= 2:
        for item in value:
            if 'кот' in item:
                answer.append(item)
    # check if the list answer contains 2 or more items break the loop
    if len(answer) >= 2:
        break
    else:
        break
            
print(answer)

['Покормить кота']


#### Множества: дубли в задачах
Последняя неделя выдалась тяжелой, и вам постоянно приходилось в спешке добавлять в записную книжку новые задачи. Теперь одна и та же задача в вашей книжке может дублироваться! Дедлайны вовсю бьют, поэтому вы готовы садиться за задачи без разбора их порядка.

Вам дан словарь с задачами `tasks`. Пройдитесь по всем ключам и устраните дублирование только в рамках значений, соответствующих одному ключу. Создайте новый словарь `new_tasks`, в котором будут те же ключи и те же задачи, но уже без дублей.

In [1]:
tasks = { 
    0: ["Покормить кота","Покормить кота"],
    1: ["Покормить кота", "Забрать посылку"]
}

# Результат
new_tasks = {
    0: ['Покормить кота'],
    1: ['Покормить кота', 'Забрать посылку']
}

Обратите внимание, что при этом:

Значения в словаре должны иметь тип list.
Порядок задач в списке не имеет значения.
NB! Переменная tasks уже загружена в ЛМС, создавать ее не нужно.

In [2]:
new_tasks = {}

for key, value in tasks.items():
    # create a new list with the unique values from the original list
    new_list = list(set(value))
    # add the key and the new list to the new dictionary
    new_tasks[key] = new_list

print(new_tasks)

{0: ['Покормить кота'], 1: ['Покормить кота', 'Забрать посылку']}


## Lesson 2: Functions, types, work with strings

### Functions

In Python, a function is a block of code that performs a specific task. Functions are used to reuse code and help break down complex problems into smaller, more manageable pieces.

To define a function in Python, you use the `def` keyword followed by the function name, and a pair of parentheses `()` containing any parameters the function might take. The code block within the function is indented and begins with a colon `:`. Here is an example of a simple function in Python:

In [3]:
def greet(name):
  print("Hello, " + name)

greet("John")  # prints "Hello, John"


Hello, John


In this example, the `greet` function takes a single parameter `name` and prints a greeting message. To call the function, you use the function `name` followed by a pair of parentheses containing the arguments you want to pass to the function.

Functions can also return a value using the `return` statement. For example:

In [4]:
def add(x, y):
  result = x + y
  return result

result = add(3, 4)  # result is 7


In this example, the `add` function takes two parameters `x` and `y`, adds them together, and returns the result. When the function is called, the value is assigned to the `result` variable.

Functions can also have default values for their parameters, which allows you to call the function with fewer arguments. For example:

In [5]:
def greet(name, greeting="Hello"):
  print(greeting + ", " + name)

greet("John")  # prints "Hello, John"
greet("John", "Hi")  # prints "Hi, John"


Hello, John
Hi, John


#### Why sometimes we need to create a variable in python and sometimes not?

In Python, you generally don't need to create variables before assigning values to them. When you assign a value to a variable, Python will automatically create the variable for you. 

In [6]:
x = 10  # creates a variable x and assigns the value 10 to it
y = "hello"  # creates a variable y and assigns the value "hello" to it


However, there are a few cases where you might want to create a variable before assigning a value to it:

1. If you want to assign a value to a variable later in your code, you can create the variable first using the `None` value:

In [7]:
x = None  # creates a variable x with a value of None

# later in the code
x = 10  # assigns the value 10 to the variable x


2. If you want to create a global variable (i.e., a variable that can be accessed from anywhere in your code), you can use the `global` keyword:

In [8]:
def foo():
  global x  # creates a global variable x
  x = 10  # assigns the value 10 to the global variable x

foo()  # calls the function foo
print(x)  # prints 10


10


3. If you want to create a constant (i.e., a variable that can't be modified), you can use all uppercase letters for the variable name and use the = operator to assign a value:

In [9]:
PI = 3.14  # creates a constant PI with a value of 3.14


#### How global variables are used in python?

In Python, a global variable is a variable that is defined outside of a function and can be accessed from anywhere in your code. Global variables are useful when you need to share a piece of information between different parts of your code, such as a user's name or a configuration setting.

In [10]:
x = 10  # creates a global variable x with a value of 10

def foo():
  global x  # tells Python that x is a global variable
  x = 20  # assigns a new value to the global variable x

foo()  # calls the function foo
print(x)  # prints 20


20


It's important to be careful when using global variables, as they can make your code harder to understand and maintain. In general, it's a good idea to minimize the use of global variables and try to use local variables whenever possible.

#### Default arguments

In Python, you can define default values for function arguments, which allows you to call the function with fewer arguments. Default arguments are specified by assigning a value to the argument in the function definition.

In [11]:
def greet(name, greeting="Hello"):
  print(greeting + ", " + name)

greet("John")  # prints "Hello, John"
greet("John", "Hi")  # prints "Hi, John"


Hello, John
Hi, John


Default arguments are useful when you want to provide a default behavior for a function, but still allow the caller to override it if needed. They can also make your code easier to read and understand by reducing the number of arguments that need to be passed to the function.

It's important to note that default arguments are evaluated only once, when the function is defined. This means that if you use a mutable object (e.g., a list or a dictionary) as a default argument, the same object will be used for all calls to the function. For example:

In [12]:
def foo(items=[]):
  items.append(1)
  print(items)

foo()  # prints [1]
foo()  # prints [1, 1]
foo()  # prints [1, 1, 1]


[1]
[1, 1]
[1, 1, 1]


#### print() function

In Python, you can use string formatting to create formatted strings that include variables and other values. String formatting allows you to insert values into a string template, which can make your code more readable and easier to maintain.

There are several ways to format strings in Python. The most common method is using the `%` operator and a format specifier. For example:

In [13]:
name = "John"
age = 30
print("%s is %d years old." % (name, age))  # prints "John is 30 years old."


John is 30 years old.


You can also use the `.format()` method to format strings. For example:

In [14]:
name = "John"
age = 30
print("{} is {} years old.".format(name, age))  # prints "John is 30 years old."


John is 30 years old.


You can also use f-strings to format strings in Python. F-strings are string literals that start with the letter `f`, and they allow you to embed expressions inside string templates. For example:

In [15]:
name = "John"
age = 30
print(f"{name} is {age} years old.")  # prints "John is 30 years old."


John is 30 years old.


#### Return statement



#### Error handling

Error handling is the process of handling runtime errors in your code. In Python, you can use the `try` and `except` statements to handle errors that might occur during the execution of your code.

In [16]:
try:
  # some code that might cause an error
  x = 10 / 0  # this line will cause a ZeroDivisionError
except ZeroDivisionError:
  # code to handle the ZeroDivisionError
  print("Cannot divide by zero!")


Cannot divide by zero!


You can also use the `finally` statement to execute code after the `try` and `except` blocks, regardless of whether an error occurred or not. For example:

In [17]:
try:
  # some code that might cause an error
  x = 10 / 0  # this line will cause a ZeroDivisionError
except ZeroDivisionError:
  # code to handle the ZeroDivisionError
  print("Cannot divide by zero!")
finally:
  # code to be executed after the try and except blocks
  print("This code will always be executed.")


Cannot divide by zero!
This code will always be executed.


In this example, the `finally` block contains code that will be executed after the `try` and `except` blocks, regardless of whether an error occurred or not.

Error handling is an important part of writing robust and reliable code. It allows you to anticipate and handle potential errors that might occur during the execution of your code, and helps prevent your program from crashing or producing unexpected results.

Error handling can also be combined with functions to create reusable error handling code. For example:

In [18]:
def get_second_element(array):
    try:
        return array[1]
    except IndexError:
        print('### Элемент не найден, печатаю все значения для разбора причин: ###')
        for elem in array:
            print(f'### {elem} ###')
        return None
    
print(f'Коррректный вызов: вернул {get_second_element((1, 2, 3))}')
print(f'Некорректный вызов: вернул {get_second_element([1])}')
# Сначала print изнутри функции, потом снаружи - ведь мы должны подсчитать get_second_element([1]), прежде чем его отдать в print()

Коррректный вызов: вернул 2
### Элемент не найден, печатаю все значения для разбора причин: ###
### 1 ###
Некорректный вызов: вернул None


#### Call stack & recursion

### Memory management in Python

#### Data model

Рассмотрим как создаются объекты в памяти, их устройство, процесс объявления новых переменных и работу операции присваивания.

Для того, чтобы объявить и сразу инициализировать переменную необходимо написать её имя, потом поставить знак равенства и значение, с которым эта переменная будет создана. Например строка:

In [ ]:
b = 5

Целочисленное значение 5 в рамках языка Python по сути своей является объектом. Объект, в данном случае – это абстракция для представления данных, данные – это числа, списки, строки и т.п. При этом, под данными следует понимать как непосредственно сами объекты, так и отношения между ними (об этом чуть позже). 
Каждый объект имеет три атрибута – это 
- идентификатор
- значение
- тип. 

Идентификатор – это уникальный признак объекта, позволяющий отличать объекты друг от друга, а значение – непосредственно информация, хранящаяся в памяти, которой управляет интерпретатор.

При инициализации переменной, на уровне интерпретатора, происходит следующее:
- создается целочисленный объект 5 (можно представить, что в этот момент создается ячейка и 5 кладется в эту ячейку);
- данный объект имеет некоторый идентификатор, значение: 5, и тип: целое число;
- посредством оператора “=” создается ссылка между переменной b и целочисленным объектом 5 (переменная b ссылается на объект 5).

Имя переменной не должно совпадать с ключевыми словами интерпретатора Python. Список ключевых слов можно найти здесь https://pythonworld.ru/osnovy/klyuchevye-slova-modul-keyword.html. Также его можно получить непосредственно в программе, для этого нужно подключить модуль keyword и воспользоваться командой keyword.kwlist.

In [19]:
import keyword
print("Python keywords: ", keyword.kwlist)

Python keywords:  ['False', 'None', 'True', 'and', 'as', 'assert', 'async', 'await', 'break', 'class', 'continue', 'def', 'del', 'elif', 'else', 'except', 'finally', 'for', 'from', 'global', 'if', 'import', 'in', 'is', 'lambda', 'nonlocal', 'not', 'or', 'pass', 'raise', 'return', 'try', 'while', 'with', 'yield']


Проверить является или нет идентификатор ключевым словом можно так:

In [21]:
keyword.iskeyword("try")

True

In [22]:
keyword.iskeyword("b")

False

Для того, чтобы посмотреть на объект с каким идентификатором ссылается данная переменная, можно использовать функцию `id()`.

In [23]:
a = 4
b = 5
id (a)

1619192185168

In [24]:
id(b)

1619192185200

In [25]:
a = b
id(a)

1619192185200

Как видно из примера, идентификатор – это некоторое целочисленное значение, посредством которого уникально адресуется объект. Изначально переменная a ссылается на объект 4 с идентификатором `1619192185168`, переменная `b` – на объект с `id = 1619192185200`. После выполнения операции присваивания `a = b`, переменная a стала ссылаться на тот же объект, что и `b`.

<img src="https://devpractice.ru/wp-content/uploads/2016/12/tpython-lesson3-1-1.png" width="400"/>

Тип переменной можно определить с помощью функции `type()`. Пример использования приведен ниже.

In [26]:
a = 10
b = "hello"
c = (1, 2)

type(a)

type(b)

type(c)

print(type(a), type(b), type(c))

<class 'int'> <class 'str'> <class 'tuple'>


##### Изменяемые и неизменяемые типы данных

В Python существуют изменяемые и неизменяемые типы.

К неизменяемым (immutable) типам относятся: целые числа (int),  числа с плавающей точкой (float), комплексные числа (complex), логические переменные (bool), кортежи (tuple), строки (str) и неизменяемые множества (frozen set).

К изменяемым (mutable) типам относятся: списки (list), множества (set), словари (dict).

Как уже было сказано ранее, при создании переменной, вначале создается объект, который имеет уникальный идентификатор, тип и значение, после этого переменная может ссылаться на созданный объект.

Неизменяемость типа данных означает, что созданный объект больше не изменяется. Например, если мы объявим переменную k = 15, то будет создан объект со значением 15, типа int и идентификатором, который можно узнать с помощью функции id().

In [28]:
k = 15
print(id(k), type(k))

1619192185520 <class 'int'>


Объект с `id = 1619192185520` будет иметь значение 15 и изменить его уже нельзя.

Если тип данных изменяемый, то можно менять значение объекта. Например, создадим список [1, 2], а потом заменим второй элемент на 3.

In [32]:
a = [1, 2]
print(id(a))

a[1] = 3
print(a)
print(id(a))



1617135692096
[1, 3]
1617135692096


Как видно, объект на который ссылается переменная a, был изменен. Это можно проиллюстрировать следующим рисунком.

<img src="https://devpractice.ru/wp-content/uploads/2016/12/python-lesson3-2-1.png" width="400"/>

В рассмотренном случае, в качестве данных списка, выступают не объекты, а отношения между объектами. Т.е. в переменной a хранятся ссылки на объекты содержащие числа 1 и 3, а не непосредственно сами эти числа.

https://devpractice.ru/python-lesson-3-data-model/